### ISSUE TITLE
Support signing in to multiple AAD directories/tenants

### ISSUE THREAD

**Issue by nikmd23** [MicrosoftDocs/vscodespaces#39]
User is seeing an issue on Twitter:

See this thread: https://twitter.com/bitandbang/status/1191370550788206592

**Comment by jpda** [comment]
Hi! I'm having the issue referenced above. This also affects my MSA, not only corp. 

Windows 10 19013, Edge Canary 80.0.320.0 and Edge Beta 79.0.309.11. Old edge 44.19013.1.0.

Opened with Guest profile, corp profile, personal profile (MSA), tried using corp and MSA to sign in. Used incognito, no extensions. The subscription list is empty and the call shows unauthorized. I sign in interactively correctly, asked for consent, etc as one would expect. Access token acquisition seems to fail due to pointing to an incorrect tenant for the token request. 

The call to /tenants returns a list of tenants, then msal goes to get a token for that tenant. In this case, even though I'm signing into the microsoft tenant (`72f988bf...`) successfully, msal is attempting to get a token for a different tenant (`98a34a88`). 

![image](https://user-images.githubusercontent.com/5932293/68148666-6faee700-ff0a-11e9-909d-0542795d4fbc.png)

The access_token used to call /tenants is issued by the home tenant with a scope of user impersonation, but /tenants returns the _other_ tenant as the first object in the list.

Not really sure why, sounds like it is a problem with the /tenants api (or maybe just my account?), but a way to choose a tenant and/or choose a default based on the value of the issuer claim in the ticket might be a decent workaround.

![image](https://user-images.githubusercontent.com/5932293/68149074-3cb92300-ff0b-11e9-93b6-bfb6021dab93.png)

**Comment by legomushroom** [comment]
Hey @jpda, thanks for the issue and the great summary 👍. That is a known bug, the reason is that we fall back to the first tenant on that list, which of course does not guarantee subscriptions. We are working on a fix at the moment which implies giving you a dropdown to choose a tenant, do you think that is a reasonable fix for you?

**Comment by legomushroom** [comment]
@nikmd23 I unassigned form myself by mistake, mind reassigning it back? For some reason, I can not do it myself.

**Comment by fara-nakms** [comment]
you can assign it to me @nikmd23

**Comment by jpda** [comment]
I think that's great @legomushroom! Even just a union of all subscriptions from all tenants would be fine, similar to how az-cli and azps do it. Anything that is discoverable. I personally like the idea of a subscription filter but any option is cool 😎

**Comment by jcormont** [comment]
Experiencing the same or a similar issue where I sign in with a personal Microsoft account, which is added to an AAD for work.

On portal.azure.com I see the correct subscriptions (and 3 AADs, including 2 old ones), but the dropdown in VS Online is only showing me an old Pay-as-you-go sub which I believe I even canceled at some point...

Can I somehow disassociate the old personal Azure account?

**Comment by jpda** [comment]
Something like this perhaps

![image](https://user-images.githubusercontent.com/5932293/68244121-216d1700-ffe2-11e9-8017-83a19846ecba.png)

demo here https://jpda.dev/aad-picker/ and code here https://github.com/jpda/aad-picker - for illustration

**Comment by colinrippeyfinarne** [comment]
Just to note that you can choose any subscription you have access to if you create a plan via the VSCode extension as the Azure Account extension will go get them for you.

**Comment by jkeech** [comment]
@colinrippeyfinarne, that's true, but when signing into VSO, it will still only use the default tenant, so plans would end up being associated with the "wrong" identity (the one from your default tenant, not the one tenant on the Azure subscription).

We are actively working on adding multi-tenant support to be able to switch directories/tenants in VSO (both the web portal and the VS Code extension), but it will probably take about a month to finish up and deploy since there are a variety of things that need to come together to support proper directory switching.

**Comment by bergmeister** [comment]
I had a even worse issue: Signed in using Tenant1 and then signed out and into Tenant2 and now I cannot use the VSO environments from Tenant2 but can still use and access the ones from Tenant1!

**Comment by eparizzi** [comment]
Any updates on this? We can't try the service because it defaults to our default directory which is not where our subscription is under and there's no way to switch.

**Comment by SylChamber** [comment]
I don't know if it is related, but I had a similar problem. I have a personal account with a VS subscription and a work account with another one. They are in different directories.

Because my organization blocks using AAD accounts for authenticating to other sites with OpenID Connect, I cannot authenticate to Codespaces using my work account. So I assigned the owner role for my work VS subscription to my personal account (it was not that easy but it's out of scope for the discussion). I can authenticate to Azure Portal with my personal account and I see and can use my work VS subscription. However Codespaces does not display my work subscription when I try to create a plan when authenticated with my personal account, even though it has ownership of my work VS subscription.

But the Visual Studio Codespaces extension for Visual Studio Code **does**. It correctly displays all subscriptions and allowed me to create a plan on my work subscription while I was authenticated with my personal account (*Codespaces: Sign in* in VS Code).

Once the plan was created, I could see it in in the Codespaces web site, and create and connect to environments. It's only the creation and deletion of plans that is problematic in the web site, they have to be done in VS Code for now. Afterwards, I can use the web site for the rest.

**Comment by legomushroom** [comment]
This is currently under development and we hope to unblock you soon.

cc @klvnraju

**Comment by jkeech** [comment]
Closing this issue as Won't Fix due to the transition to GitHub Codespaces which uses GitHub authentication.
